joblib/numpy_pickle.py

```diff
@@ -307,25 +314,23 @@ def __init__(self, filename, file_handle, mmap_mode=None):
             np = None
         self.np = np
 
-    def _open_pickle(self, file_handle):
-        return file_handle
-
     def load_build(self):
-        """ This method is called to set the state of a newly created
-            object.
+        """Called to set the state of a newly created object.
 
-            We capture it to replace our place-holder objects,
-            NDArrayWrapper, by the array we are interested in. We
-            replace them directly in the stack of pickler.
+        We capture it to replace our place-holder objects,
+        NDArrayWrapper, by the array we are interested in. We
+        replace them directly in the stack of pickler.
         """
         Unpickler.load_build(self)
-        if isinstance(self.stack[-1], NDArrayWrapper):
+
+        # For back backward compatibility
+        if isinstance(self.stack[-1], (NDArrayWrapper, NumpyArrayWrapper)):
             if self.np is None:
-                raise ImportError('Trying to unpickle an ndarray, '
-                        "but numpy didn't import correctly")
-            nd_array_wrapper = self.stack.pop()
-            array = nd_array_wrapper.read(self)
-            self.stack.append(array)
+                raise ImportError("Trying to unpickle an ndarray, "
+                                  "but numpy didn't import correctly")
+            array_wrapper = self.stack.pop()
+            self.compat_mode = isinstance(array_wrapper, NDArrayWrapper)
```

Comment:
Quite unlikely but you could imagine a mix of NDArrayWrapper and NumpyArrayWrapper in a pickle and it feels like this would unpickle fine. I guess you need to have self.compat_mode set to True if one of the wrapper is a NDArrayWrapper. At the moment, you use the last wrapper class instead.